crates/storage/src/collections/binary_heap/children_vec.rs

```diff
@@ -0,0 +1,460 @@
+    /// of yielded elements.
+    pub fn iter(&self) -> Iter<T> {
+        Iter::new(&self)
+    }
+
+    /// Returns an iterator yielding exclusive references to all elements of the heap.
+    ///
+    /// # Note
+    ///
+    /// Avoid unbounded iteration over big storage heaps.
+    /// Prefer using methods like `Iterator::take` in order to limit the number
+    /// of yielded elements.
+    pub fn iter_mut(&mut self) -> IterMut<T> {
+        IterMut::new(self)
+    }
+
+    /// Returns a shared reference to the first element if any.
+    pub fn first(&self) -> Option<&T> {
+        if self.is_empty() {
+            return None
+        }
+        self.get(0)
+    }
+
+    /// Returns an exclusive reference to the first element if any.
+    pub fn first_mut(&mut self) -> Option<&mut T> {
+        if self.is_empty() {
+            return None
+        }
+        self.get_mut(0)
+    }
+
+    /// Removes all elements from this heap.
+    ///
+    /// # Note
+    ///
+    /// Use this method to clear the heap instead of e.g. iterative `pop()`.
+    /// This method performs significantly better and does not actually read
+    /// any of the elements (whereas `pop()` does).
+    pub fn clear(&mut self) {
+        if self.is_empty() {
+            return
+        }
+        self.children.clear();
+        self.len = Lazy::new(0);
+    }
+
+    /// Appends an element to the back of the heap.
+    pub fn push(&mut self, value: T) {
+        assert!(
+            self.len() < core::u32::MAX,
+            "cannot push more elements into the storage heap"
+        );
+        let last_index = self.len();
+        *self.len += 1;
+        self.push_to(last_index, Some(value));
+    }
+
+    /// Returns information about the child at the heap index if any.
+    pub(super) fn get_child(&self, index: u32) -> Option<ChildInfo<T>> {
+        let storage_index = children::get_children_storage_index(index);
+        let child_pos = children::get_child_pos(index);
+        let children = self.children.get(storage_index)?;
+        let child = children.child(child_pos);
+        Some(ChildInfo::new(child))
+    }
+
+    /// Returns information about the child at the heap index if any.
+    ///
+    /// The returned `ChildInfoMut` contains a mutable reference to the value `T`.
+    pub(super) fn get_child_mut(&mut self, index: u32) -> Option<ChildInfoMut<T>> {
+        let storage_index = children::get_children_storage_index(index);
+        let child_pos = children::get_child_pos(index);
+        let children = self.children.get_mut(storage_index)?;
+        let count = children.count();
+        let child = children.child_mut(child_pos);
+        Some(ChildInfoMut::new(child, count))
+    }
+
+    /// Pushes `value` to the heap index `index`.
+    ///
+    /// If there is already a child in storage which `index` resolves to
+    /// then `value` is inserted there. Otherwise a new child is created.
+    fn push_to(&mut self, index: u32, value: Option<T>) {
```

Comment:
Normally this is called `insert`.

Reply:
So there is no need for a symmetric `remove` operation?

Reply:
Nope, since the only operation necessary for removing is `pop` (also for balancing that's the only operation necessary).

Reply:
Ah and I renamed it to `insert`, yeah that works better. It was kind of a legacy naming.